### PR TITLE
Quote environment paths so ones with spaces do not break activation

### PR DESCRIPTION
**Pull request by notluquis**
Refs #837, deliberately not `Fixes`: the reported scenario still fails elsewhere, see the end.

`createCommandScriptInEnv` builds a shell script as a string and `runCommandInEnvironment` hands it to `bash -c` or `cmd /c`. The interpolated paths were quoted wrongly on both platforms.

```ts
scriptLines.push(`CALL ${activatePath}`);                              // windows: not quoted at all
scriptLines.push(`source ${quoteChar}${activatePath}${quoteChar}`);    // posix: double quotes
```

`CALL C:/Users/First Last/mambaforge/Scripts/activate` splits at the space, which is #837. Double quotes prevent word-splitting but still evaluate `$(...)` and backticks, which is the CodeQL finding.

`shellQuotePath(value, isWin)` now single-quotes with the `'\''` escape on posix and double-quotes on `cmd`, which has no command substitution. Applied to the activate path, the `conda.sh` path, the env path and the `-p` argument. The `quoteChar` option is gone: both callers passed `"'"` only so their output would survive `launchTerminalInDirectory`, and single quotes are the default now.

Ran a prefix containing both a space and `$(id -u)` through bash:

```
new: source '/tmp/.../First Last/$(id -u)/bin/activate'   ->  ACTIVATED_OK
old: source "/tmp/.../First Last/$(id -u)/bin/activate"   ->  No such file or directory
```

Five tests added; reverting the quoting fails three of them.

What this does not close:

- **#837 itself.** `createLaunchScript` in `server.ts` still emits `CALL ${envActivatePath}`, `CALL conda activate ${envPath}` and `source ${condaShellScriptPath}` unquoted, and every session launch goes through it. #1076 has the full map.
- **The CodeQL alerts on `env.ts`.** Not in this diff; the taint path from `env.ts:337` into the spawn sinks is intact.
- **Windows.** I cannot test it. The unit test pins the generated string, but someone should reproduce #837 on a real box before treating it as closed.
- **The AppleScript escape chains** at `utils.ts:746` and `:864-869`, one of which runs with administrator privileges. I round-tripped quotes, backslashes, backticks, `$()` and newlines through real `osascript` and all came back byte-exact, but deleting either chain leaves the suite green, so a test for them is a fair review request.

Worked through this with Claude Code; the bash output above is a run, not a description of one.

## AI usage

- **YES**: Some or all of the content of this PR was generated by AI.
- ****: The human author has carefully reviewed this PR and run this code (keep this PR "draft" until the answer is YES)
- AI tools and models used: Claude Code, Opus 5

### CODE REVIEW

**Comment by notluquis** [comment]
Pushed a second commit. The first one was incomplete and the tests I added made it look finished.

Quoting inside `createCommandScriptInEnv` is correct for the `env.ts` path, which spawns `bash -c` with the script as its own argv entry. It does nothing for the two `pythonenvdialog` callers, because `launchTerminalInDirectory` composed that script into another string and ran it through `exec()`, which spawns `/bin/sh`. The outer shell expanded `$(...)` before bash saw the single quotes.

So the fix is not more quoting, it is not handing shells strings at all:

```ts
exec(`open "${dirPath}"`)            ->  execFile('open', [dirPath])
exec(`gnome-terminal ... -- bash -c "${commands}"`)
                                     ->  execFile('gnome-terminal', ['--working-directory=' + dirPath, '--', 'bash', '-c', commands])
```

Same for the `osascript` calls including `setupJlabCLICommandWithElevatedRights`, which runs with administrator privileges and was building its command by interpolation. Where a shell is genuinely needed the paths are quoted instead: `createUnsignScriptInEnv` and `conda-unpack`, whose Windows branch was unquoted, which is #837 again in the installer path.

Three test problems from the same review are fixed too: the Windows assertion was a prefix match that omitted the closing quote, so removing that quote still passed; the conda `-p` quoting had no coverage; and nothing covered the launcher argv. All three now fail when the corresponding code is reverted.

File: src/main/utils.ts [hunk 1]
```diff
@@ -755,18 +776,21 @@ export function launchTerminalInDirectory(options: {
...
+        '--',
+        'bash',
+        '-c',
+        `${commands}${interactive ? '; exec bash' : ''}`
+      );
     }
-    exec(`gnome-terminal --working-directory="${dirPath}"${callCommands}`);
+    execFile('gnome-terminal', args);
   }
 }
```
[the remark this answers is not in the document]
Comment (krassowski) [reply src/main/utils.ts:789]: Pre-existing

**Comment by notluquis** [comment]
Applied: `launchTerminalInDirectory` now returns `true` on the success path instead of falling off the end. Neither caller reads the result, so behaviour is unchanged; the declared type just was not true. Two tests pin both returns. Mutation check I ran locally: removing the new `return` fails the success one and leaves the other 89 in the file passing. Written with Claude Code.

**Review by krassowski — COMMENTED** [review]
Feedback from Opus 5 session regarding Windows (I do not have access to a Windows machine to verify at the moment):

> - The Windows terminal launch without commands passes the directory as a bare argv element, and cmd.exe re-parses it. src/main/utils.ts:771 builds cmd /c start cmd.exe /K cd /D <dirPath>, and libuv only adds quotes when an argument contains a space, tab or double quote, so &, ^, |, < and > reach cmd unquoted. A folder named R&D opens the terminal in the parent directory and runs D as a second command; master's cd /D "..." kept the & literal, so this is a regression. Passing the whole string with windowsVerbatimArguments: true restores it, while pre-quoting the argv element does not, because libuv rewrites the embedded quotes as \".
> - shellQuotePath does not double a trailing backslash on Windows, so the value can escape its own closing quote. src/main/utils.ts:542 returns "D:\" for a drive root, and the caller at pythonenvdialog.ts:171 emits jupyter lab --notebook-dir="D:\". CommandLineToArgvW reads \" as a literal quote, so jupyter_server receives D:" and raises a TraitError. Master emitted 'D:\', which traitlets stripped, so the drive-root case regresses while the much more common space case is fixed; the fix is to double the run of backslashes immediately before the closing quote.

It would be useful to expand test coverage so that our CI can test all edge cases on real Windows.

**Comment by notluquis** [comment]
Ran the Windows check this says it could not do, on `windows-latest` (Server 2025). Eleven directory names, each legal on NTFS, each holding an `activate.bat` that echoes a marker, driven through `spawn('cmd', ['/c', script], { windowsVerbatimArguments: true })` in the shape `runCommandInEnvironment` uses.

#837 is fixed. `First Last` returns the marker quoted and fails unquoted, which is the whole claim.

| name | quoted | unquoted |
| --- | --- | --- |
| `First Last` | ok | fails |
| `a&b` | ok | fails |
| `a(b)c` | ok | fails |
| `a;b` | ok | fails |
| `a=b` | ok | fails |
| `it's` | ok | ok |
| `100%` | ok | ok |
| `%NOPE_UNDEFINED%` | ok | ok |
| `%USERNAME%` | **fails** | fails |
| `a^b` | **fails** | fails |
| `!PATH!` | ok, **fails** with DelayedExpansion | ok |

So the quoting carries five names that were broken without it, and three stay broken with it.

`%USERNAME%` and `!PATH!` are cmd expansion, which happens inside double quotes rather than despite them. Microsoft's [cmd reference](https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/cmd) puts it plainly, with no exception for quoting: "To substitute variable values in the command line or scripts, enclose the variable name in percent signs (%VariableName%)." `!` needs `DelayedExpansion` set in `HKCU\Software\Microsoft\Command Processor`, which is off by default; the run above turns it on to show the difference. An undefined name survives, which is why `%NOPE_UNDEFINED%` passes and `%USERNAME%` does not. That makes the comment at `utils.ts:542` the part that needs adjusting: cmd does substitute, just not the `$(...)` kind.

`a^b` is not `CALL`. Same path without `CALL` fails the same way. Reading it against the same page, `^` is in the list of characters that void the quote-preservation rule for `/c`, so the outer quotes are stripped and `^` is back to being the escape character. That is inference from the doc rather than something isolated further.

Worth knowing before anyone reaches for it: dropping `CALL` is worse, not better. `cmd /c "<path>"` with no `CALL` breaks `a&b`, `a(b)c`, `a;b` and `a=b` as well, for the same quote-stripping reason.

Nothing here ran anything of a caller's choosing. Each failure lands on a path that does not exist and `CALL` reports that. Not tested: an environment variable whose value itself contains a quote.

Worked through this with Claude Code. The table is a run on a hosted Windows runner rather than a reading of the code, and the `CALL` line is there because I predicted the opposite and the run disagreed.

**Comment by notluquis** [comment]
Went back for the case the previous comment left open, an environment variable whose value carries a quote. It is not contained.

Reachability first, because it is the part that decides how much this matters. Two things have to line up: a directory in the environment path has to be named after a variable, in the `%NAME%` form, and that variable has to be set to a value containing a double quote. cmd expands the name inside the quotes `shellQuotePath` adds, the quote in the value closes them, and the rest of the value is parsed as command text rather than as part of a path. Anyone able to set a variable in the session that launches the app can already run code as that user by other means, so this does not look like a boundary being crossed, and the run showed the injected text reaching a command rather than anything about privilege. Worth a second opinion on that reading rather than my word for it.

Not posting the strings. Happy to send them privately if it helps review.

A fix does exist, and it is measured rather than proposed. `%` has no escape on the cmd command line, but `%%` is one inside a batch file, so writing the `CALL` into a temporary `.bat` and doubling the percent signs in the path contains every case while still reaching the target. Two details that only showed up by running it:

- doubling once and keeping `CALL` contains the injection but never reaches the target either, including for a plain harmless value. Containment by breaking, which is easy to mistake for a fix when the only thing checked is that nothing bad ran.
- `CALL` expands a second time, so it needs four percent signs where a plain invocation inside the same batch file needs two. Both of those reach the target and contain all four values.

None of this changes the merge call as far as I can see. Every case in the table from the previous comment is the same or better with this change than without it, and #837 is fixed. What it does change is the claim the code makes about itself: the Windows branch is an improvement, not a safe quote, and the comment at `utils.ts:542` is where that should be said.

Worked through this with Claude Code, on a hosted Windows runner. The first candidate fix is in the list above because I would have shipped it if I had stopped at "nothing bad ran".

**Comment by notluquis** [comment]
One more of the same family, found by sweeping for the shape rather than by reading this diff.

`setupJlabCommandWithUserRights` in `src/main/utils.ts` builds a shell command by interpolation and runs it:

```
const cmd = `ln -s ${targetPath} ${symlinkPath}`;
execSync(cmd, { shell: '/bin/bash' });
```

`targetPath` is `${getAppDir()}/app/jlab`, which on macOS sits inside the `.app` bundle. Nothing stops that path from holding a space: the bundle can be renamed, and it can be installed somewhere like `~/Downloads/My Apps/`. `ln` then gets three arguments and fails, and the CLI setup reports nothing useful. It is the same class this pull request is about, just outside `createCommandScriptInEnv`.

`fs.symlinkSync(targetPath, symlinkPath)` does the same job with no shell to quote for, which is a smaller fix than quoting. This runs on the non-elevated path, so a permission failure lands in the `try` that is already there. The elevated path nearby builds its command differently and would need its own look.

Also in the same function, `fs.chmodSync(symlinkPath, 0o755)` follows the link, so it chmods the target rather than the link, and the very next line chmods the target again. Harmless, but one of the two lines is not doing what its name suggests.

Not opening this as a change here, since it is outside the diff. Happy to take it in this pull request if that is the preference, or file it separately.

Worked through this with Claude Code, while checking whether the symlink and quoting mistakes it found elsewhere in the config work had siblings.

**Comment by krassowski** [comment]
It looks like it picked up a merge conflict due to the comment rewrite (I do not mind it being one way or another).
